### PR TITLE
Automatic content syncing: preferred instance tracking and fixes

### LINKED ISSUE #10732: InsufficientStorage error raised after running a full sync with KDP
## Observed behavior

After running a full sync with KDP, the InsufficientStorage error was raised from the function `_process_content_requests`. The message `Content download requests need 0B of free space` was received which suggests that the user does not have enough storage, despite free_space being `15061171584`

Possibly caused by this [while statement](https://github.com/learningequality/kolibri/blob/4cf6a66db0714f76735d2b695d4291276b7b2d4b/kolibri/core/content/utils/content_request.py#LL404C2-L404C2) or [download_request assignment](https://github.com/learningequality/kolibri/blob/4cf6a66db0714f76735d2b695d4291276b7b2d4b/kolibri/core/content/utils/content_request.py#L419) within `content_request.py`

Relevant:
- It was the first time this facility had synced to KDP
- Registration was skipped
- Lessons and quizzes were in the local facility data (the lessons & quizzes were created, and the resources used were then deleted from the device)
## Errors and logs

```
[python-devserver] DEBUG    2023-05-21 17:07:53,388 Did not find suitable download request for free space 15061171584
[python-devserver] WARNING  2023-05-21 17:07:53,392 Content download requests need 0B of free space
[python-devserver] ERROR    2023-05-21 17:07:53,393 Job ca69893aff45483e8bc769daf078c825 raised an exception: Traceback (most recent call last):
[python-devserver]   File "/home/richard/github/kolibri/kolibri/core/content/utils/content_request.py", line 266, in process_content_requests
[python-devserver]     _process_content_requests(incomplete_downloads)
[python-devserver]   File "/home/richard/github/kolibri/kolibri/core/content/utils/content_request.py", line 445, in _process_content_requests
[python-devserver]     raise InsufficientStorage(
[python-devserver] kolibri.core.content.utils.content_request.InsufficientStorage: Content download requests need 0B of free space
```

## Expected behavior

If there is enough free space, the `download_request` should be able to be processed.
## User-facing consequences

User devices unable to process download requests despite having enough space to do so.

### LINKED ISSUE #11022: User initiated content download needs remote browsing context (for Studio)
## Observed behavior

When browsing Studio content remotely through Kolibri, if I initiate a download of one of the resources, it fails to 'find a peer' for importing the content.

## Errors and logs
```
INFO     2023-07-27 10:22:17,150 No acceptable peer network device for importing content metadata
WARNING  2023-07-27 10:22:17,160 Could not find available peer for content import
```

## Expected behavior

When a user initiates a download of a resource through remote browsing, it should store the device on which the user was browsing, so that the automatic import processing can choose the best device. This will be primarily useful for the use case of browsing Studio, but is valuable information to have in any case for optimizing the import processing.

## User-facing consequences

Content fails to download and remains in 'Waiting to download' state

## Steps to reproduce

1) Provision Kolibri
2) Browse local libraries, navigating to Studio
3) Find a resource and click the download button to initiate the download
4) Observe the download never completes

## Context

Kolibri 0.16.0b0
https://www.notion.so/learningequality/Initiating-a-content-download-from-Studio-runs-into-No-acceptable-peer-network-device-for-importing-306e6139324145e18d78057e2683ad35?pvs=4
https://www.notion.so/learningequality/Initiating-a-content-download-from-a-peer-that-has-the-content-failed-8ddc95fe6742464da9e4ff26b3049694?pvs=4

### PR DESCRIPTION
**Pull request by bjester**
## Summary
Resolves issues with stalled 'My Downloads' through final integration of device settings features and usage of a reserved Network Location for Studio, allowing us to set a preference for which device to source downloads from.

### Details

- Adds `source_instance_id` to `ContentRequest` model as the preferred instance to source the content from
- Updates `ContentDownloadRequest` API to support `source_instance_id`
- Adds validator to automatic content download task, aligned with whether or not it's enabled
- Adds tests for `content_request` utilities handling metadata import, peer processing, and total size annotations
- Corrects issues with total size annotations returning `0`
- Filters out incomplete downloads for learners when learners are disallowed from creating download requests
- Updates `NetworkLocation` models to use `location_type` field instead of `dynamic` boolean and necessary migrations (planned in https://github.com/learningequality/kolibri/issues/10431)
- Creates `reserved` `NetworkLocation` for Studio using UUID generated from `CENTRAL_CONTENT_BASE_URL`
- Updates `NetworkLocation` API to allow detail retrieval using `instance_id` assuming lack of conflicts, and excludes reserved locations in list responses
- Prevents strict comparison of `instance_id` and overwriting it for `reserved` `NetworkLocation`s until we can get Studio to return the same instance ID
- Consolidates duplicate `NetworkLocationResource` APIs on the frontend
- Refactors existing `canDownload` allowing users to download resources externally, to disambiguate between new download feature
- Updates Learn `useDevices` composable to compare route and current device by `instance_id`
- Adds logging to `catch()` in learn's routing, and removes potentially duplicate calls to `router.replace`
- Integrates `allow_learner_download_resources` device setting in the frontend to only show the download button when enabled
- Upgrades `le_utils` from `0.1.38` to `0.2.1`

## References

https://github.com/learningequality/le-utils/pull/111
Fixes https://github.com/learningequality/kolibri/issues/10732
Resolves https://github.com/learningequality/kolibri/issues/11022
Contributes to https://github.com/learningequality/kolibri/issues/10386
Starts https://github.com/learningequality/kolibri/issues/10431

## Reviewer guidance

1) Requires you to manually run:
```
kolibri manage migrate discovery --database networklocation
```
2) In order to download from Studio, you need to change the setting in `options.ini` to the hotfixes server
----

## Testing checklist

- [X] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [x] Critical and brittle code paths are covered by unit tests

## PR process

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

File: kolibri/core/content/test/test_content_app.py
Comment (rtibbles) [inline kolibri/core/content/test/test_content_app.py]: I think you can also just set this as a class level property:

```
maxDiff = 10000
```

or can set to `None` to give the whole output.

File: kolibri/core/content/utils/content_request.py
Comment (rtibbles) [inline kolibri/core/content/utils/content_request.py:388]: Is the error this is fixing specific to using the Exists subquery in a `then` statement?

File: kolibri/core/content/utils/content_request.py
Comment (rtibbles) [inline kolibri/core/content/utils/content_request.py]: Shouldn't we be filtering by Kolibri version here?

File: kolibri/core/content/utils/content_request.py
Comment (rtibbles) [inline kolibri/core/content/utils/content_request.py]: Could we return the number of successfully imported content nodes from this function, so we can break from the loop once we've finished importing everything?

Alternatively, we can evaluate the unprocessed count after each import instead of outside of the loop.

File: kolibri/core/content/utils/content_request.py
Comment (rtibbles) [inline kolibri/core/content/utils/content_request.py]: I wonder if the _get_preferred_network_location function could be implemented slightly differently to accept the full list of source_instance_ids and return a generator that can then be iterated below `for peer in peers` and we just keep going through the list of peers generated by that until all the requested resources have been imported.

This could also then include the 'trusted' peer behaviour even in the case where source_instance_ids have been passed too.

File: kolibri/plugins/learn/assets/src/composables/useDevices.js
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/composables/useDevices.js:21]: If we didn't filter the Reserved locations out of the NetworkLocations API response, could we just be getting most of this data from the API instead now? (we'd still need to translate the name, and filter it for appropriate learner types).

**Review by rtibbles — COMMENTED** [review]
Just a couple of questions - but nothing that should be blocking merge.

File: kolibri/core/content/test/test_content_app.py
Comment (bjester) [reply kolibri/core/content/test/test_content_app.py]: ah! sounds good. I thought it peculiar that it said `self.maxDiff` explicitly, but I'll do this

File: kolibri/core/content/utils/content_request.py
Comment (bjester) [reply kolibri/core/content/utils/content_request.py:388]: I'm not sure. There seemed to be some mixing of parameter definitions and invocation styles with regards to single args and keyword args in the calls to `resolve_expression`. The function parameters are indeed all keyword args, but I recall seeing instances in other Django expressions where it was called with plain args, like this `Case(then=)` does here, but they happened to match the keyword arguments, which would still work right? The `Exists` wildcarded the keyword args, which seemed to cause the issue.

File: kolibri/core/content/utils/content_request.py
Comment (bjester) [reply kolibri/core/content/utils/content_request.py]: Definitely! :facepalm:

File: kolibri/core/content/utils/content_request.py
Comment (bjester) [reply kolibri/core/content/utils/content_request.py]: Yes :+1:

File: kolibri/core/content/utils/content_request.py
Comment (bjester) [reply kolibri/core/content/utils/content_request.py]: I think that's a good idea and would make it more flexible for other usage. I'll implement that

File: kolibri/plugins/learn/assets/src/composables/useDevices.js
Comment (bjester) [reply kolibri/plugins/learn/assets/src/composables/useDevices.js:21]: Yes, very likely. I was unsure of the implications of preventing the reserved Studio instance from the connection testing, and how that would be handled by consuming usages, so I decided to leave that to the scope of the other outstanding issue

File: kolibri/core/content/utils/content_request.py
Comment (bjester) [reply kolibri/core/content/utils/content_request.py]: Oh wait, no! This is just the file import, not the metadata import, so it should be okay to leave unfiltered?

Comment (rtibbles) [reply kolibri/core/content/utils/content_request.py]: Ahhhh, ok!

Comment (rtibbles) [reply kolibri/core/content/utils/content_request.py]: Maybe add a comment for future people thinking they are being helpful like me?

Comment (bjester) [reply kolibri/core/content/utils/content_request.py]: Yep, will do!

**Review by rtibbles — APPROVED** [review]
Code changes make sense, and manual testing checks out (weird local data issues aside!)

**Pull request merged**
